Add the idl files and stubs for User Agent Client Hints

colelao opened the pull request:
#### d521e56de4b718ab0be019f5214abfeb2ee56429
<pre>
Add the idl files and stubs for User Agent Client Hints
<a href="https://bugs.webkit.org/show_bug.cgi?id=296355">https://bugs.webkit.org/show_bug.cgi?id=296355</a>
<a href="https://rdar.apple.com/156450572">rdar://156450572</a>

Reviewed by Brent Fulgham.

This change adds the idl files for the User Agent Client Hints JavaScript API as outlined here:
<a href="https://wicg.github.io/ua-client-hints/#idl-index">https://wicg.github.io/ua-client-hints/#idl-index</a>

It also adds the initial stubs for the UserAgentData data structure.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/page/Navigator.cpp:
(WebCore::Navigator::initializeNavigatorUAData const):
(WebCore::Navigator::userAgentData const):
* Source/WebCore/page/Navigator.h:
* Source/WebCore/page/Navigator.idl:
* Source/WebCore/page/NavigatorUA.idl: Added.
* Source/WebCore/page/NavigatorUABrandVersion.h: Added.
* Source/WebCore/page/NavigatorUABrandVersion.idl: Added.
* Source/WebCore/page/NavigatorUAData.cpp: Added.
(WebCore::NavigatorUAData::create):
(WebCore::NavigatorUAData::brands const):
(WebCore::NavigatorUAData::mobile const):
(WebCore::NavigatorUAData::platform const):
(WebCore::NavigatorUAData::toJSON const):
(WebCore::NavigatorUAData::getHighEntropyValues const):
* Source/WebCore/page/NavigatorUAData.h: Added.
* Source/WebCore/page/NavigatorUAData.idl: Added.
* Source/WebCore/page/PointerLockController.cpp:
* Source/WebCore/page/UADataValues.h: Added.
* Source/WebCore/page/UADataValues.idl: Added.
* Source/WebCore/page/UALowEntropyJSON.h: Added.
* Source/WebCore/page/UALowEntropyJSON.idl: Added.
* Source/WebCore/page/WorkerNavigator.cpp:
(WebCore::WorkerNavigator::initializeNavigatorUAData const):
(WebCore::WorkerNavigator::userAgentData const):
* Source/WebCore/page/WorkerNavigator.h:
* Source/WebCore/page/WorkerNavigator.idl:
* Source/WebCore/page/ios/ContentChangeObserver.cpp:

Canonical link: <a href="https://commits.webkit.org/297899@main">https://commits.webkit.org/297899@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bafb48895f267ae2aa96f6a47c6aaad25f193bc6

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/113234 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/32974 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/23447 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/119442 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/63993 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/115196 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/33622 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/41532 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/86183 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/41257 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/116181 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/26860 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/101862 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/66505 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/26131 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/19990 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/63196 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/96241 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/20065 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/122659 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/40312 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/30075 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/95031 "Passed tests") | 
| | [❌ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/40697 "Hash bafb4889 for PR 48399 does not build (failure)") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/98077 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/94773 "Passed tests") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/17737 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/36448 "The change is no longer eligible for processing.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18207 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/40198 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/45697 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/39839 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/43172 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/41576 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WebCore/DerivedSources.make:
```diff
@@ -1551,6 +1551,11 @@ JS_BINDING_IDLS := \
     $(WebCore)/page/NavigatorStorage.idl \
+    $(WebCore)/page/NavigatorUA.idl \
+    $(WebCore)/page/NavigatorUAData.idl \
+    $(WebCore)/page/NavigatorUABrandVersion.idl \
```
Nit: `NavigatorUABrandVersion.idl` should precede `NavigatorUAData.idl`

brentfulgham commented on Source/WebCore/Sources.txt:
```diff
@@ -4289,20 +4290,25 @@ JSMutationEvent.cpp
 JSMutationObserver.cpp
 JSMutationRecord.cpp
 JSNamedNodeMap.cpp
+JSCustomXPathNSResolver.cpp
```
Nit: Why did this item move from alphabetical position? That seems like a mistake.

brentfulgham replied:
... or else it should move far earlier in the file!

brentfulgham commented on Source/WebCore/page/Navigator.h:
```diff
@@ -84,6 +86,7 @@ class Navigator final
     explicit Navigator(ScriptExecutionContext*, LocalDOMWindow&);
 
     void initializePluginAndMimeTypeArrays();
+    void initializeNavigatorUAData() const;
```
It's surprising that an initialization function for this class would be `const`. What can it possibly be doing that doesn't modify the state of the object?

colelao replied:
It should modify only the `mutable` variable `m_navigatorUAData`, so `const` should be okay. This allows the `userAgentData()` function to remain `const` while only initializing the data structure when it is requested.

brentfulgham commented on Source/WebCore/page/Navigator.cpp:
```diff
@@ -447,4 +448,19 @@ int Navigator::maxTouchPoints() const
+    if (m_navigatorUAData)
+        return;
+
+    // TODO: populate the data structure
```
Nit: We file a bugzilla for tasks like this (e.g., "populate new NavigatorUAData structures"), and then write this as `FIXME(123456): Populate the data structure.`

Please file that Bugzilla (and import to Radar) and revise this comment.

brentfulgham commented on Source/WebCore/page/NavigatorUAData.h:
```diff
@@ -0,0 +1,54 @@
+
+#include "IDLTypes.h"
+#include "JSDOMPromiseDeferred.h"
+#include "NavigatorUABrandVersion.h"
```
I suspect we can forward-declare `NavigatorUABrandVersion.h` and avoid the header include here.

brentfulgham commented on Source/WebCore/page/NavigatorUAData.h:
```diff
@@ -0,0 +1,54 @@
+#include "JSDOMPromiseDeferred.h"
+#include "NavigatorUABrandVersion.h"
+#include "UADataValues.h"
+#include "UALowEntropyJSON.h"
```
Ditto these two includes. Try to forward-declare instead.

brentfulgham commented on Source/WebCore/page/NavigatorUAData.h:
```diff
@@ -0,0 +1,54 @@
+#include "UADataValues.h"
+#include "UALowEntropyJSON.h"
+#include <wtf/Vector.h>
+#include <wtf/text/WTFString.h>
```
Include `<wtf/Forward.h>` instead.

brentfulgham commented on Source/WebCore/page/NavigatorUAData.h:
```diff
@@ -0,0 +1,54 @@
+    static Ref<NavigatorUAData> create();
+    const Vector<NavigatorUABrandVersion>& brands() const;
+    bool mobile() const;
+    String platform() const;
```
Does `platform` need to be returned by copy?

brentfulgham commented on Source/WebCore/page/NavigatorUAData.h:
```diff
@@ -0,0 +1,54 @@
+    UALowEntropyJSON toJSON() const;
+
+    using ValuesPromise = DOMPromiseDeferred<IDLSequence<IDLDictionary<UADataValues>>>;
+    void getHighEntropyValues(Vector<String> hints, ValuesPromise&&) const;
```
Do we need to pass an expensive `Vector<String>` here by copy? Seems like a const ref would work.

brentfulgham commented on Source/WebCore/page/PointerLockController.cpp:
```diff
@@ -42,6 +42,7 @@
 #include "VoidCallback.h"
 #include <wtf/Ref.h>
 #include <wtf/TZoneMallocInlines.h>
+#include <JavaScriptCore/ConsoleTypes.h>
```
Was this needed because the new files moved unified builds around triggering a compile error?

colelao replied:
Yes, I was getting build errors without it.

brentfulgham commented on Source/WebCore/page/WorkerNavigator.cpp:
```diff
@@ -109,4 +110,19 @@ void WorkerNavigator::clearAppBadge(Ref<DeferredPromise>&& promise)
+}
+
+NavigatorUAData& WorkerNavigator::userAgentData() const
+{
```
Probably better as:

```
if (!m_navigatorUAData)
    initalizeNavigatorUAData();

return *m_navigatorUAData;
```

brentfulgham reviewed (changes requested):
This is looking pretty good, but I think it needs a few adjustments before we can land it.

brentfulgham commented on Source/WebCore/DerivedSources.make:
```diff
@@ -58,6 +58,33 @@ to-pattern = $(join $(basename $1), $(subst .,%,$(suffix $1)))
+	$(WebCore)/Modules/indexeddb/IDBTransactionDurability.idl \
+	$(WebCore)/css/typedom/CSSKeywordValue.idl \
+	$(WebCore)/css/typedom/numeric/CSSMathNegate.idl \
+	$(WebCore)/css/typedom/transform/CSSMatrixComponent.idl \
```
Something weird happened here. Did you text editor decide to do some kind of whitespace cleanup? Please revert that so we only have meaningful changes in this file.

colelao replied:
I just reordered the lines with an editor function; It looks like the indentation was already present. I will fix this in the follow up patch

brentfulgham commented on Source/WebCore/Sources.txt:
```diff
@@ -218,15 +218,15 @@ Modules/indexeddb/shared/IndexKey.cpp
 Modules/mediacapabilities/NavigatorMediaCapabilities.cpp
 Modules/mediacapabilities/WorkerNavigatorMediaCapabilities.cpp
+Modules/mediacontrols/MediaControlsHost.cpp
+Modules/mediarecorder/BlobEvent.cpp
```
Nit: If you are going to do some re-alphabetizing, be complete! :-) BlobEvent should precede MediaControlsHost.

brentfulgham replied:
Actually, this is a huge reordering of this file. I suggest you revert all of this and just do the change needed for his patch. Then follow-up with a simple "reorder" patch that does nothing but put this file in the proper order.

colelao replied:
Yeah, no problem. That totally makes sense!

brentfulgham replied:
It's probably totally fine, but I've been bitten by small mistakes in stuff like this, so it's nice to do them in a single patch that can easily be reverted if needed. Sorry for the make-work! :-)

brentfulgham commented on Source/WebCore/page/NavigatorUAData.cpp:
```diff
@@ -0,0 +1,69 @@
+}
+
+void NavigatorUAData::getHighEntropyValues(const Vector<String>&, NavigatorUAData::ValuesPromise&&) const
+{
```
You can also use `notImplemented();` in here to output a log message but probably not needed for this project since you are going to fill this in soon.

brentfulgham reviewed (changes requested):
I think this patch looks great, but I'm uncomfortable landing the large file-alphebetizing changes with this patch. Please revert the large-scale reordering and just include the bits needed to implement this feature. Then, file a new bug to land the full reordering. This will make things easier to review, and easier for us to revert this feature on its own if we want to at some point.

brentfulgham reviewed (approved):
Thank you for cleaning up those small issues. R=me

colelao committed:
Add the idl files and stubs for User Agent Client Hints
https://bugs.webkit.org/show_bug.cgi?id=296355
rdar://156450572

Reviewed by Brent Fulgham.

This change adds the idl files for the User Agent Client Hints JavaScript API as outlined here:
https://wicg.github.io/ua-client-hints/#idl-index

It also adds the initial stubs for the UserAgentData data structure.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/page/Navigator.cpp:
(WebCore::Navigator::initializeNavigatorUAData const):
(WebCore::Navigator::userAgentData const):
* Source/WebCore/page/Navigator.h:
* Source/WebCore/page/Navigator.idl:
* Source/WebCore/page/NavigatorUA.idl: Added.
* Source/WebCore/page/NavigatorUABrandVersion.h: Added.
* Source/WebCore/page/NavigatorUABrandVersion.idl: Added.
* Source/WebCore/page/NavigatorUAData.cpp: Added.
(WebCore::NavigatorUAData::create):
(WebCore::NavigatorUAData::brands const):
(WebCore::NavigatorUAData::mobile const):
(WebCore::NavigatorUAData::platform const):
(WebCore::NavigatorUAData::toJSON const):
(WebCore::NavigatorUAData::getHighEntropyValues const):
* Source/WebCore/page/NavigatorUAData.h: Added.
* Source/WebCore/page/NavigatorUAData.idl: Added.
* Source/WebCore/page/PointerLockController.cpp:
* Source/WebCore/page/UADataValues.h: Added.
* Source/WebCore/page/UADataValues.idl: Added.
* Source/WebCore/page/UALowEntropyJSON.h: Added.
* Source/WebCore/page/UALowEntropyJSON.idl: Added.
* Source/WebCore/page/WorkerNavigator.cpp:
(WebCore::WorkerNavigator::initializeNavigatorUAData const):
(WebCore::WorkerNavigator::userAgentData const):
* Source/WebCore/page/WorkerNavigator.h:
* Source/WebCore/page/WorkerNavigator.idl:
* Source/WebCore/page/ios/ContentChangeObserver.cpp:

Canonical link: https://commits.webkit.org/297899@main
